### ISSUE TITLE
Put server_capability status code is incorrect

### ISSUE THREAD

**Issue by ZugNZwang** [apache/trafficcontrol#4000]
## I'm submitting a ...

- [X] bug report
- [ ] new feature / enhancement request
- [ ] improvement request (usability, performance, tech debt, etc.)
- [ ] other 

## Traffic Control components affected ...

- [ ] CDN in a Box
- [ ] Documentation
- [ ] Grove
- [ ] Traffic Control Client
- [ ] Traffic Monitor
- [X] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] unknown

## Current behavior:

When HTTP PUT https://{{TO_BASE_URL}}/api/{{api_version}}/server_capabilities is made the status code received is incorrect. Receiving 404.
## Expected / new behavior:

A 405 status code is expected.
## Anything else:

on latest master - last commit 3.0.0-10254.f1d15647.el7

**Comment by ocket8888** [comment]
This is a consequence of how routing is done. I know I've brought this up before, but since I can't find an issue for it already, we'll make this the canonical one.

Our API improperly treats a request method as part of a unique resource identifier - this is pervasive throughout the TO API on all endpoints (i.e. not just server_capability).

**Comment by mitchell852** [comment]
afaik our api never returns 405s so 405 would not be "right". 404 seems "right" to me if you `PUT https://{{TO_BASE_URL}}/api/{{api_version}}/server_capabilities` without a query param then the resource to PUT cannot be found.

**Comment by ocket8888** [comment]
What this really comes down as far as path parameters go is a willingness to reveal the existence of other method handlers. For example, you can't give back a 405 for `PATCH /deliveryservices/xmlid-that-my-tenant-does-not-have-access-to` because that discloses the existence of the delivery service (and also you'd need to make database calls for a request you already know is erroneous and can't accomplish anything), or you could just do what we do and return a 404.

This is part of why I like query parameters better for keys. If somebody does `PATCH /deliveryservices?xmlid=same-as-above`, you can both safely and correctly give back a 405, because the parameter is no longer part of the resource path.

**Comment by mitchell852** [comment]
sorry, i'm a little slow on the uptake here. `Put`. I thought PUT was actually implemented and the bug was a 404 caused by the lack of query params...

Not sure what the behavior should be. 405 seems right. or i even thought we did 501 at one point.

**Comment by mitchell852** [comment]
also, this seems `minor` to me as an informed api consumer would not be calling PUT anyhow but that's just my opinion.

**Comment by ocket8888** [comment]
> _" I thought PUT was actually implemented and the bug was a 404 caused by the lack of query params"_

No, the error happens because there is no `PUT` handler; our routing considers request method to be a part of a URI.

> _"this seems minor to me"_

Well, its impact is minuscule, but its scope is massive. It affects every single endpoint in the TO API (afaik), so I figured medium was appropriate. But I don't feel strongly about it.

**Comment by rob05c** [comment]
It's not "incorrect," it's just less precise than it could be. The HTTP spec (RFC 7231) doesn't require a 405, you can always use a less precise code, and 404 is correct, the resource does not exist there.

It wouldn't be a bad thing to change TO to return 405's, but I wouldn't call it a bug.

**Comment by mitchell852** [comment]
> I just labeled it as it was reported, shall I change it?

yeah, maybe @ZugNZwang can change his description to a `improvement request` as this is more of an API improvement rather than a bug and then we can change the label...

**Comment by mitchell852** [comment]
@ZugNZwang - can you retitle this / rewrite the description a bit to be more generic?
